49 currency system

Heldereim opened the pull request:
Added a currency system to the game. Player starts with 300 currency which is reduced when placing buildings or as building upkeep when running. Currency can be gained by selling the stored resources of a building or setting the building itself to sell its production as soon as possible. 

Branch also comes with a few QOL changes. When scrolling in UI elements the game map no longer gets zoomed in/out.

Heldereim committed:
Selectables scripts folder

Created a new folder in scripts called selectables and put all the selectables scripts in it

Heldereim committed:
Started player currency work

Started working on player currency. Added a global script as well as an initial scene housing the panel that will display the players currency amount.

Heldereim committed:
Icon placeholder

Added a placeholder image to the icon in the player currency panel.

Heldereim committed:
Added icons, added coin animation

Heldereim committed:
Player currency panel, Build select list bug fix

Fixed an issue with the build select list not opening when clicked; its mouse filter was set to stop, this was then changed to pass. Player currency panel is now part of user interface

Heldereim committed:
Buildmanager logic for player currency

Added logic hindering the placement of a building if the player can't afford it as well as removing the currency from the player equal to the cost of the building when placing a building.

Heldereim committed:
Work done for currency panel

Added a working animation to the currency panel for the rotating coin. Removed the building cost field from selectable buildings and added them to the building data base class. Added visitor pattern in the BuildingInfo class working together with Building data class and its inheritors. Split parts of a spritesheet into individual images to use in the animation. Created a sprite frames resource from a sprite atlas. Set the building cost of some of the buildings resources. Set mouse filter stop for most of the UI scenes

Heldereim committed:
Broke up duplicated code

Broke out parts of text formating code in Buildinginfo and made them into separate functions

Heldereim committed:
Merge branch 'main' into 49-currency-system

Heldereim committed:
BuildingInfo panel text formating

After the merge from main, added specific text formating in the buildinginfo panel to account for area gathering buildings. Also added a pollutions text formatter. Made it so that more complex buildings reuse the text formating of the less complex buildings. Added a unique version of the accept function to areagathering building data

Heldereim committed:
Removed debug prints

Heldereim committed:
Added upkeep

Added an upkeep variable to building data

Heldereim committed:
Storage tab building info

There is now a second tab in the building info panel. It has no functionality yet but should be used to view the storage of a building and sell off parts of that storage.

Heldereim committed:
BuildingInfo TabBar

Heldereim committed:
Merge branch 'main' into 49-currency-system

Heldereim committed:
Resource costs

Added resource costs to the enums script as well as a function to access them.

Heldereim committed:
Temporary sell button, scripts

Made a temporary no styled button for selling production. Added parts in the script in BuildingInfo for disabling and enabling the sell production button depending on the building type. Added a current building variable in the BuildingInfo script to keep track of the currently targeted building. Added code in the scripts for both gathering building and production building since they have different functions for producing goods.

Heldereim committed:
Script changes, selling production

Changed the way buildings handle their production so that production can be sold in a middle step. Made an extra step in building info script specific to storage buildings. Added a button to change buildings to start storing their production (or outputting it). Added accept function in storage building data script.

Heldereim committed:
BuildingInfo styling

Added some theme styling to the Store/Sell buttons in building info panel. Added a theme to the tab bars for when they are disabled.

Heldereim committed:
buildinginfo changes

Changed the font color on the building info sell/store buttons. Set the building upkeep for the coal mine, iron mine, and wood cutter. Set the building cost for the wood cutter at 40. Added code in output resources part of the prod building script checking whether the player has the required currency for production. Increased starting currency to 300 in PlayerCurrency script.

Heldereim committed:
PlayerCurrency Script changes

Removed an unused function and variable. Added comments

Heldereim committed:
Build list scale reset

Set the building select list scale to 1 in its scene and in the UI scene

Heldereim committed:
Stored resources

Added a few styleboxtexture resources. Added a stored resource scene that is to be used in the building info panel. Also added a script for this scene. Working on script for populating the stored_resource panels.

Heldereim committed:
Scripts updated

Building info panel now continuously updates the held resources number in all its stored resources panels

Heldereim committed:
Building info panel additions

Added a button for selling the chosen stored resources. Also added a panel displaying the amount of currency to be gained if the chosen amounts are sold.

Heldereim committed:
Zooming, mouse filter, ui signals, ui element

Added a signal in user interface that is emitted when the mouse enters any ui element. Created a script called ui element from which all ui main scenes inherit. Updated a bunch of mouse filter properties on different ui elements. Added code in build manager that hinders building a factory if the mouse is over a ui element. Building info panel, building select list, player currency panel, resource ui and the panel in user interface now extend the ui element class. User interface now keeps a list of all ui elements that have a mouse hovering over them.

Heldereim committed:
ui style, continuous production sell stored resources

Added the _process function to production building so that buildings continually check whether they can produce resources. This solves the issue with buildings not producing resources if they have reached their max

Heldereim committed:
Bugfix

Fixed a bug where the label showing the value of chosen resources would not update when pressing on a new building

Heldereim committed:
Merge branch 'main' into 49-currency-system

linuskar reviewed (changes requested):
Cost and upkeep missing for:
- Biomass power plant
- Coal power plant
- Warehouse, probably no upkeep here just cost?
- Landfill, probably no upkeep here just cost?
- Road, probably no upkeep here just cost?

One thing that would be helpful is seeing what selling method is currently selected for a building. 
![image](https://github.com/user-attachments/assets/d6f6acca-a89e-460b-a0e9-7257468d369f)

linuskar reviewed (changes requested):
Can't see storage for biomass in relevant buildings like warehouse, biomass landfill and woodcutter.

linuskar commented on scripts/utility/camera_movement.gd:
```diff
@@ -33,19 +36,21 @@ func _process(delta):
+				zoomTarget *= 1 + zoomSensitivity
+			else:
+				zoomTarget = Vector2(maxZoom, maxZoom)
+			print(zoomTarget)
```
Remove debug print

Heldereim replied:
Removed the print

linuskar commented on scripts/UI/player_currency_panel.gd:
```diff
@@ -0,0 +1,6 @@
+extends UIElement
+
+@onready var player_currency = $Styling/HBoxContainer/MarginContainer/CurrencyAmount
```
Type hinting

Heldereim replied:
Added ': Label' after player_currency

linuskar commented on scripts/utility/map_areas.gd:
What does the signal for allowing zoom do when entering map bounds here?

Heldereim replied:
Old attempt at solving the problem which proved unreliable, will remove

linuskar commented on scripts/utility/enums.gd:
```diff
@@ -62,7 +69,22 @@ static var tile_names: Dictionary[TileType, String] = {
 	TileType.RESOURCE: "RESOURCE",
-	}	
+	}
+# List of the value of every resource when sold
```
Add resource cost for biomass or will it default to 0?

Heldereim replied:
Since biomass is not something we were planning on letting the player sell i figured it was ok to not have a cost associated with it.

linuskar commented on scripts/utility/enums.gd:
```diff
@@ -15,7 +15,14 @@ static var emissions: Array[ResourceType] = [ResourceType.CO2, ResourceType.S02]
 	ResourceType.COAL, ResourceType.ELECTRICITY, ResourceType.WOOD]
-	
+
+static var resource_image_paths: Dictionary[ResourceType, String] = {
```
Image path for biomass

Heldereim replied:
Added image path for bio mass

Heldereim committed:
Changes made for merge

Added upkeep and/or cost for all buildings. Removed old attempt at solving zoom issue. Added type hints in building info script. Added a label showing what mode a building was set to currently (store/sell). Removed debug print.

linuskar commented on resources/buildings/road.tres:
```diff
@@ -7,4 +7,6 @@ script = ExtResource("1_ob7tx")
 valid_tile_types_to_place_on = Array[int]([1, 2, 3])
 building_size = Vector2(1, 1)
+building_cost = 10
+building_upkeep = 10
```
Should there be upkeep for road?

Heldereim replied:
No, fixed now

linuskar commented on scripts/buildings/building_data/building_data.gd:
```diff
@@ -12,3 +12,10 @@ extends Resource
+@export var building_cost: int
+## The upkeep cost of the building
+@export var building_upkeep: int
+
```
Maybe add some description of what the function does. Hard to know what it does at first look

Heldereim replied:
Added a description/use case but this is part of the visitor pattern implementation and is hard enough to understand even if you know how it works.

Heldereim committed:
storage for byproducts, filepath for biomass image

Added function for hiding the buttons in a stored resource panel. Added a filepath for the biomass image in enums. Added specific scenarios in building info panel for when a resource is a byproduct. Added a comment in accept function of building data explaining the way to use the function. Removed upkeep from road.

Heldereim committed:
More fixes for merge

linuskar commented on scripts/UI/selectables/stored_resource.gd:
```diff
@@ -0,0 +1,91 @@
+
+## Hides the parts of the stored resource related to selling that resource
+func disable_selling() -> void:
+	$PanelContainer/MarginContainer/HBoxContainer/MarginContainer/SellingStorage/Decrease.hide()
```
Maybe better to have these as instance variables and call hide function on the variables for better readability here.

Heldereim replied:
Probably correct

linuskar commented on scripts/UI/selectables/stored_resource.gd:
```diff
@@ -0,0 +1,91 @@
+## selling resources by 5
+func decrease_more() -> void:
+	if resource_to_sell - 5 <= 0:
+		
```
whitespace, line 73

Heldereim replied:
removed

Heldereim committed:
More fixes for merge

Heldereim committed:
Misspelling

linuskar commented on scripts/UI/BuildingInfo.gd:
```diff
@@ -1,72 +1,208 @@
-@onready var image = $ScrollContainer/VBoxContainer/MarginContainer3/FactoryImage
-@onready var building_name = $ScrollContainer/VBoxContainer/BuildingName
-@onready var info = $ScrollContainer/VBoxContainer/MarginContainer4/PanelContainer/MarginContainer/BuildingInfo
+@onready var image = $MarginContainer/General/VBoxContainer/MarginContainer3/FactoryImage
```
Missing type hinting, image, building_name, info

Heldereim replied:
Fixed

linuskar reviewed (changes requested):
Would be good to be able to see the upkeep cost in the building select list UI and in the UI for selecting a building, general tab.

Heldereim committed:
Changes made

Heldereim committed:
Updated clickable hit box for 2x2 buildings

linuskar commented on scripts/UI/BuildingInfo.gd:
```diff
@@ -1,72 +1,214 @@
+	current_building = building
 	self.show()
 	populate_info_label(building)
+	if current_building is StorageBuilding:
```
When trying to select a building like warehouse or landfill then it crashes for trying to set building for selling. Can use the function is_a_warehouse from Enums with the current building to not set selling then?

linuskar replied:
I did a solution like this and it worked
![image](https://github.com/user-attachments/assets/009cc01c-ebac-48c0-b2a8-28c8a560c232)

linuskar commented:
Also there seems to be a problem in the UI with selling multiple resources from a building. The functionality works in selling and getting currency but the UI does not update.

In the warehouse the proper amount that is sold is shown in the UI for wood, but not the other resources.
![image](https://github.com/user-attachments/assets/2c871b0d-917e-4efc-ad72-fa183944cc8e)
![image](https://github.com/user-attachments/assets/1d50ccc8-1d2e-43b5-a545-c0125c4c80cf)
![image](https://github.com/user-attachments/assets/707a04af-5c03-4a29-a4df-55ef7162710f)

Heldereim committed:
Change for bugfix

Heldereim committed:
Update for bugfix

Heldereim committed:
Scrollable selectables

Updates made to fix a feature in a pull request

Heldereim committed:
Bugfixes for merge

Fixed a bug where resources were not accurately removed from warehouses when sold through the buildinginfo menu. Fixed a bug where the selling tab in the building info panel could be accessed on a warehouse by first pressing a different building, selecting the selling tab and then pressing the warehouse. This would then lead to a crash if the player pressed a selling mode.

Heldereim committed:
Removed unused method

linuskar reviewed (approved):
LGTM
